File: frontend-next/src/lib/weather-api.ts
/**
 * Typed client for the KrishiAI weather backend.
 *
 * All requests go through the Next.js rewrite (/api/weather/*  ->  :8000/api/weather/*).
 * Every function throws on non-2xx with a friendly message so callers can just
 * catch and surface `err.message` in the UI.
 */

// -------- Types (mirror backend/schemas/weather.py) --------

export interface Location {
  name: string
  region?: string | null
  country?: string | null
  lat: number
  lon: number
  timezone?: string | null
}

export interface LocationHit {
  name: string
  region?: string | null
  country?: string | null
  lat: number
  lon: number
  display_name: string
}

export interface CurrentWeather {
  temp_c: number
  temp_f: number
  feels_like_c: number
  feels_like_f: number
  humidity: number
  pressure_mb?: number | null
  wind_kph: number
  wind_mph: number
  wind_dir?: string | null
  wind_deg?: number | null
  condition: string
  condition_code?: string | null
  icon?: string | null
  visibility_km?: number | null
  uv_index?: number | null
  cloud_cover?: number | null
  is_day?: boolean | null
  observed_at?: string | null
}

export interface UnifiedWeather {
  location: Location
  current: CurrentWeather
  provider: string
  cached: boolean
  fetched_at: string
}

export interface HourPoint {
  time: string
  temp_c: number
  temp_f: number
  condition: string
  icon?: string | null
  humidity?: number | null
  wind_kph?: number | null
  chance_of_rain?: number | null
}

export interface DayPoint {
  date: string
  temp_min_c: number
  temp_max_c: number
  temp_min_f: number
  temp_max_f: number
  condition: string
  icon?: string | null
  humidity?: number | null
  wind_kph?: number | null
  chance_of_rain?: number | null
  sunrise?: string | null
  sunset?: string | null
}

export interface ForecastBundle {
  location: Location
  hourly: HourPoint[]
  daily: DayPoint[]
  provider: string
  cached: boolean
  fetched_at: string
}

export interface GeoIPResult {
  ip?: string
  lat: number
  lon: number
  city?: string
  region?: string
  country?: string
  country_code?: string
  timezone?: string
  fallback?: boolean
}

export interface ProviderHealth {
  name: string
  configured: boolean
  priority: number
}

export interface HealthResponse {
  providers: ProviderHealth[]
  cache_backend: string
  rate_limit: string
}

// -------- Fetch helper --------

class WeatherAPIError extends Error {
  status: number
  constructor(message: string, status: number) {
    super(message)
    this.name = "WeatherAPIError"
    this.status = status
  }
}

async function apiGet<T>(path: string, signal?: AbortSignal): Promise<T> {
  const res = await fetch(path, { signal, cache: "no-store" })
  if (!res.ok) {
    let detail = `Request failed (${res.status})`
    try {
      const j = await res.json()
      if (j?.detail) detail = typeof j.detail === "string" ? j.detail : JSON.stringify(j.detail)
    } catch {
      /* ignore parse errors */
    }
    throw new WeatherAPIError(detail, res.status)
  }
  return res.json() as Promise<T>
}

// -------- Public API --------

export function getCurrentByCoords(
  lat: number,
  lon: number,
  signal?: AbortSignal
): Promise<UnifiedWeather> {
  return apiGet<UnifiedWeather>(`/api/weather/current?lat=${lat}&lon=${lon}`, signal)
}

export function getCurrentByCity(q: string, signal?: AbortSignal): Promise<UnifiedWeather> {
  return apiGet<UnifiedWeather>(`/api/weather/current?q=${encodeURIComponent(q)}`, signal)
}

export function getForecastByCoords(
  lat: number,
  lon: number,
  days = 5,
  signal?: AbortSignal
): Promise<ForecastBundle> {
  return apiGet<ForecastBundle>(
    `/api/weather/forecast?lat=${lat}&lon=${lon}&days=${days}`,
    signal
  )
}

export function searchLocations(
  q: string,
  limit = 5,
  signal?: AbortSignal
): Promise<{ results: LocationHit[]; query: string }> {
  return apiGet<{ results: LocationHit[]; query: string }>(
    `/api/weather/search?q=${encodeURIComponent(q)}&limit=${limit}`,
    signal
  )
}

export function geoipLookup(signal?: AbortSignal): Promise<GeoIPResult> {
  return apiGet<GeoIPResult>(`/api/weather/geoip`, signal)
}

export function getWeatherHealth(signal?: AbortSignal): Promise<HealthResponse> {
  return apiGet<HealthResponse>(`/api/weather/health`, signal)
}

// -------- Browser geolocation (with IP fallback) --------

export interface ResolvedLocation {
  lat: number
  lon: number
  source: "browser" | "ip" | "default"
  city?: string
  country?: string
}

/**
 * Resolves the user's current location using this chain:
 *   1. navigator.geolocation (if granted within `timeoutMs`)
 *   2. /api/weather/geoip
 *   3. Hardcoded Delhi fallback
 *
 * Never throws — always returns a location.
 */
export async function resolveUserLocation(timeoutMs = 4000): Promise<ResolvedLocation> {
  // Try the browser first.
  if (typeof navigator !== "undefined" && navigator.geolocation) {
    const browserLoc = await new Promise<ResolvedLocation | null>((resolve) => {
      const timer = setTimeout(() => resolve(null), timeoutMs)
      navigator.geolocation.getCurrentPosition(
        (pos) => {
          clearTimeout(timer)
          resolve({
            lat: pos.coords.latitude,
            lon: pos.coords.longitude,
            source: "browser",
          })
        },
        () => {
          clearTimeout(timer)
          resolve(null)
        },
        { enableHighAccuracy: false, timeout: timeoutMs, maximumAge: 5 * 60 * 1000 }
      )
    })
    if (browserLoc) return browserLoc
  }

  // IP-based fallback.
  try {
    const info = await geoipLookup()
    return {
      lat: info.lat,
      lon: info.lon,
      city: info.city,
      country: info.country,
      source: info.fallback ? "default" : "ip",
    }
  } catch {
    return { lat: 28.6139, lon: 77.209, city: "Delhi", country: "India", source: "default" }
  }
}

export { WeatherAPIError }

File: backend/services/providers/weatherapi.py
"""WeatherAPI.com provider — fallback.

Docs: https://www.weatherapi.com/docs/
Free tier: 1M calls/month, includes forecast + search.
"""
from __future__ import annotations

from datetime import datetime, timezone
from typing import List

from schemas.weather import (
    CurrentWeather,
    DayPoint,
    ForecastBundle,
    HourPoint,
    Location,
    LocationHit,
    UnifiedWeather,
)
from .base import WeatherProvider


_BASE = "https://api.weatherapi.com/v1"


def _c_to_f(c: float) -> float:
    return round(c * 9 / 5 + 32, 1)


class WeatherAPIProvider(WeatherProvider):
    name = "weatherapi"
    priority = 20

    async def fetch_current(self, lat: float, lon: float) -> UnifiedWeather:
        data = await self._get(
            f"{_BASE}/current.json",
            params={"key": self.api_key, "q": f"{lat},{lon}", "aqi": "no"},
        )

        loc_data = data.get("location", {})
        cur = data.get("current", {})
        cond = cur.get("condition", {}) or {}

        loc = Location(
            name=loc_data.get("name", "Unknown"),
            region=loc_data.get("region"),
            country=loc_data.get("country"),
            lat=float(loc_data.get("lat", lat)),
            lon=float(loc_data.get("lon", lon)),
            timezone=loc_data.get("tz_id"),
        )
        current = CurrentWeather(
            temp_c=float(cur.get("temp_c", 0)),
            temp_f=float(cur.get("temp_f", 0)),
            feels_like_c=float(cur.get("feelslike_c", 0)),
            feels_like_f=float(cur.get("feelslike_f", 0)),
            humidity=int(cur.get("humidity", 0)),
            pressure_mb=float(cur.get("pressure_mb", 0) or 0),
            wind_kph=float(cur.get("wind_kph", 0) or 0),
            wind_mph=float(cur.get("wind_mph", 0) or 0),
            wind_dir=cur.get("wind_dir"),
            wind_deg=float(cur.get("wind_degree", 0) or 0),
            condition=cond.get("text", ""),
            condition_code=str(cond.get("code", "")),
            icon=cond.get("icon"),
            visibility_km=float(cur.get("vis_km", 0) or 0),
            uv_index=float(cur.get("uv", 0) or 0),
            cloud_cover=int(cur.get("cloud", 0) or 0),
            is_day=bool(cur.get("is_day", 1)),
            observed_at=datetime.fromtimestamp(cur.get("last_updated_epoch", 0), tz=timezone.utc)
                if cur.get("last_updated_epoch") else None,
        )
        return UnifiedWeather(
            location=loc,
            current=current,
            provider=self.name,
            cached=False,
            fetched_at=datetime.now(timezone.utc),
        )

    async def fetch_forecast(self, lat: float, lon: float, days: int = 5) -> ForecastBundle:
        data = await self._get(
            f"{_BASE}/forecast.json",
            params={"key": self.api_key, "q": f"{lat},{lon}", "days": min(10, days), "aqi": "no", "alerts": "no"},
        )

        loc_data = data.get("location", {})
        loc = Location(
            name=loc_data.get("name", "Unknown"),
            region=loc_data.get("region"),
            country=loc_data.get("country"),
            lat=float(loc_data.get("lat", lat)),
            lon=float(loc_data.get("lon", lon)),
            timezone=loc_data.get("tz_id"),
        )

        hourly: List[HourPoint] = []
        daily: List[DayPoint] = []

        for day in (data.get("forecast", {}).get("forecastday") or []):
            d = day.get("day", {})
            astro = day.get("astro", {})
            cond = d.get("condition", {}) or {}
            daily.append(DayPoint(
                date=day.get("date", ""),
                temp_min_c=float(d.get("mintemp_c", 0)),
                temp_max_c=float(d.get("maxtemp_c", 0)),
                temp_min_f=float(d.get("mintemp_f", 0)),
                temp_max_f=float(d.get("maxtemp_f", 0)),
                condition=cond.get("text", ""),
                icon=cond.get("icon"),
                humidity=int(d.get("avghumidity", 0)),
                wind_kph=float(d.get("maxwind_kph", 0) or 0),
                chance_of_rain=int(d.get("daily_chance_of_rain", 0) or 0),
                sunrise=astro.get("sunrise"),
                sunset=astro.get("sunset"),
            ))
            for h in day.get("hour", []):
                hcond = h.get("condition", {}) or {}
                hourly.append(HourPoint(
                    time=datetime.fromtimestamp(h.get("time_epoch", 0), tz=timezone.utc),
                    temp_c=float(h.get("temp_c", 0)),
                    temp_f=float(h.get("temp_f", 0)),
                    condition=hcond.get("text", ""),
                    icon=hcond.get("icon"),
                    humidity=int(h.get("humidity", 0)),
                    wind_kph=float(h.get("wind_kph", 0) or 0),
                    chance_of_rain=int(h.get("chance_of_rain", 0) or 0),
                ))

        return ForecastBundle(
            location=loc,
            hourly=hourly[:24],
            daily=daily[:days],
            provider=self.name,
            cached=False,
            fetched_at=datetime.now(timezone.utc),
        )

    async def search_locations(self, query: str, limit: int = 5) -> List[LocationHit]:
        data = await self._get(
            f"{_BASE}/search.json",
            params={"key": self.api_key, "q": query},
        )
        if not isinstance(data, list):
            return []

        hits: List[LocationHit] = []
        for item in data[:limit]:
            parts = [item.get("name"), item.get("region"), item.get("country")]
            parts = [p for p in parts if p]
            hits.append(LocationHit(
                name=item.get("name", ""),
                region=item.get("region"),
                country=item.get("country"),
                lat=float(item["lat"]),
                lon=float(item["lon"]),
                display_name=", ".join(parts),
            ))
        return hits
